src/bitcoin-cli.cpp

```diff
@@ -474,6 +515,48 @@ static int CommandLineRPC(int argc, char *argv[])
         if (gArgs.GetBoolArg("-getinfo", false)) {
             rh.reset(new GetinfoRequestHandler());
             method = "";
+        } else if (gArgs.GetBoolArg("-generate", false)) {
+            //-generate is special as it executes two RPC calls:
+            // 1) "getnewaddress" without any parameters, which returns a bech32 address
+            // 2) "generatetoaddress" with three params: numblocks, address and maxtries
+            // If no args are given the args-vector will be extended with DEFAULT_GENERATED_NBLOCKS
+            // and DEFAULT_GENERATED_MAXTRIES.
```

Comment:
Why is is needed to extend them with the defaults?

Reply:
In generatetoaddress the first and third param are **nblocks* and **maxtries**.  This is to make sure that all three args are there. But if there is a better way to handle it, please, give me a hint.

Reply:
If the user does not give a value, you can just pass nothing along as well (or null). The server then checks this and fills in the current default: https://github.com/bitcoin/bitcoin/blob/a739d207a39fc34c2e5adf4e14bfbb3a6288ede1/src/rpc/mining.cpp#L209

Reply:
But I need ti to construct the "fake" RPC call, because I also have to generate an address first. Then I go to create blocks. So I have to "save" all those params and their positions, so I can find them again.

Reply:
But of course, I could simply put a NULL there, no problem. Let me check it how it'll behave. :)

Reply:
Hm, I'm now getting errors. No matter if I enter any args or none at all. Not sure if I should redo the logic? Is it really important that it **doesn't use** any default args?

Reply:
So I tried this diff:

```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index ef05c910d4..2fff98c05c 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -535,11 +535,11 @@ static int CommandLineRPC(int argc, char *argv[])
                 if (args.size() == 0) {
                     args.push_back(std::to_string(DEFAULT_GENERATED_NBLOCKS));
                     args.push_back(std::move(address));
-                    args.push_back(std::to_string(DEFAULT_GENERATED_MAXTRIES));
                 }
                 if (args.size() < 2) {
                     args.push_back(std::move(address));
-                    args.push_back(std::to_string(DEFAULT_GENERATED_MAXTRIES));
                 }
                 if (args.size() == 2) {
                     const std::string nblocks = args[0];
```

And I get

```
$ ./src/bitcoin-cli -generate 1 
error: Error parsing JSON:bcrt1qyqfgcaddym58h6py4rrsz4g428gl5tmw84rt36
```

Not sure why that is. I might take another look later.

Reply:
This is surely because of the "ephemeral" calls in between. First, the address gets generated, then its JSON result flows into the second call, generatetoaddress. I am manipulating the args vector, which is being (re)used by the two APIs.

Reply:
Oh, I forgot to replace the `if` with `else if`. This works:

```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index ef05c910d4..4179520cc1 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -229,7 +229,7 @@ class GetNewAddressRequestHandler: public BaseRequestHandler
 public:
    const int ID_GETNEWADDRESS = 0;
 
-   UniValue PrepareRequest(const std::string& method, const std::vector<std::string>& args) override
+   UniValue PrepareRequest(const std::string& method, const std::vector<std::string>& /* args */) override
     {
         const UniValue& request = JSONRPCRequestObj("getnewaddress", NullUniValue, ID_GETNEWADDRESS);
         return request;
@@ -248,11 +248,8 @@ public:
 
     UniValue PrepareRequest(const std::string& method, const std::vector<std::string>& args) override
     {
-        std::vector<std::string> params = {};
-        params.push_back(args[0]);
-        params.push_back(args[1]);
-        params.push_back(args[2]);
-        UniValue paramsObj = RPCConvertValues("generatetoaddress", params);
+        UniValue paramsObj = RPCConvertValues("generatetoaddress", args);
         return JSONRPCRequestObj("generatetoaddress", paramsObj, ID_GENERATETOADDRESS);
     }
 
@@ -535,13 +532,11 @@ static int CommandLineRPC(int argc, char *argv[])
                 if (args.size() == 0) {
                     args.push_back(std::to_string(DEFAULT_GENERATED_NBLOCKS));
                     args.push_back(std::move(address));
-                    args.push_back(std::to_string(DEFAULT_GENERATED_MAXTRIES));
                 }
-                if (args.size() < 2) {
+                else if (args.size() < 2) {
                     args.push_back(std::move(address));
-                    args.push_back(std::to_string(DEFAULT_GENERATED_MAXTRIES));
                 }
-                if (args.size() == 2) {
+                else if (args.size() == 2) {
                     const std::string nblocks = args[0];
                     const std::string maxtries = args[1];
                     args.clear();

Reply:
Great! Many thanks!